fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReplaceSubqueryTypeRewriteRule.java

```diff
@@ -0,0 +1,98 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.rewrite.scalar;
+
+import com.google.common.collect.Lists;
+import com.starrocks.catalog.Function;
+import com.starrocks.catalog.FunctionSet;
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.catalog.Type;
+import com.starrocks.sql.common.TypeManager;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CaseWhenOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ScalarOperatorRewriteContext;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * The processing flow of the subquery is as follows:
+ * 1. Transform `Expr` expression to `ScalarOperator` expression, while subquery in simply wrapped in `SubqueryOperator`
+ * 2. Parse `SubqueryOperator` to apply operator, and save the `SubqueryOperator -> ColumnRefOperator` in ExpressionMapping
+ * 3. rewrite `ScalarOperator` created by step1 with the `ExpressionMapping` created by step2
+ * <p>
+ * Some function, such as `if(cond, subquery1, subquery2)`, it's output type depends on the arg types,
+ * but during step1, we cannot derive the real output type of subquery1 or subquery2 until they are parsed
+ * So this rule, executed in step3, will recalculate the argType or returnType of such kind of functions
+ */
+public class ReplaceSubqueryTypeRewriteRule extends BottomUpScalarOperatorRewriteRule {
+
+    @Override
+    public ScalarOperator visitCaseWhenOperator(CaseWhenOperator operator, ScalarOperatorRewriteContext context) {
+        List<Type> types = Lists.newArrayList();
+        for (int i = 0; i < operator.getWhenClauseSize(); i++) {
+            types.add(operator.getThenClause(i).getType());
+        }
+        if (operator.hasElse()) {
+            types.add(operator.getElseClause().getType());
+        }
+
+        Type returnType = TypeManager.getCompatibleTypeForCaseWhen(types);
+        if (Objects.equals(returnType, operator.getType())) {
+            return operator;
+        }
+
+        return new CaseWhenOperator(returnType, operator);
```

Comment:
can we directly set the returnType use setType() method here?

Reply:
Emm, better to create a new Object if changed.  And the rewriter control flow will determine whether changed by simply comparing the obj reference

```java
    private ScalarOperator applyRuleBottomUp(ScalarOperator operator, ScalarOperatorRewriteRule rule) {
        for (int i = 0; i < operator.getChildren().size(); i++) {
            operator.setChild(i, applyRuleBottomUp(operator.getChild(i), rule));
        }

        ScalarOperator op = rule.apply(operator, context);
        if (op != operator) {
            context.change();
        }
        return op;
    }
```

Reply:
ok